File: python/location_shortlist.py
from __future__ import annotations

import os
from typing import Any, Dict, List

from nexez_agent_sdk import NexezApiError, create_client


def main() -> None:
    nexez = create_client(buyer_agent="nexez-location-shortlist-example")

    buyer_intent = os.getenv("NEXEZ_BUYER_INTENT", "find a remote AI workflow consultant under 3000")
    buyer_location = os.getenv("NEXEZ_BUYER_LOCATION", "Chicago, IL")
    buyer_budget = os.getenv("NEXEZ_BUYER_BUDGET", "USD 2500")

    matches = nexez.search(buyer_intent, location=buyer_location, limit=8)
    candidates: List[Dict[str, Any]] = []

    unique_results = []
    seen_slugs = set()
    for result in matches.get("results", []):
        slug = result["page"]["slug"]
        if slug in seen_slugs:
            continue
        seen_slugs.add(slug)
        unique_results.append(result)
    for result in unique_results[:5]:
        manifest = nexez.get_agent_page(result["page"]["slug"])
        offer_key = (result.get("offer") or {}).get("key") or (manifest.get("offers") or [{}])[0].get("key")
        offer = next((item for item in manifest.get("offers", []) if offer_key and item.get("key") == offer_key), None)

        candidates.append(
            {
                "result": result,
                "manifest": manifest,
                "offer": offer,
                "score": score_candidate(result, manifest),
            }
        )

    shortlist = sorted([item for item in candidates if item.get("offer")], key=lambda item: item["score"], reverse=True)[:3]

    print(
        {
            "query": buyer_intent,
            "location": buyer_location,
            "shortlist": [
                {
                    "score": item["score"],
                    "page": item["result"]["page"]["name"],
                    "slug": item["result"]["page"]["slug"],
                    "seller_location": item["result"]["page"].get("location") or item["manifest"]["page"].get("location"),
                    "offer": item["offer"].get("name"),
                    "price": item["offer"].get("price"),
                    "agent_json": item["result"]["page"]["agent_json_url"],
                }
                for item in shortlist
            ],
        }
    )

    if not shortlist:
        print("No actionable Nexez page found for this buyer intent.")
        return

    top = shortlist[0]
    offer = top["offer"]
    supports_negotiation = bool(offer.get("negotiation_action"))

    if supports_negotiation:
        validation = nexez.validate_negotiation(
            slug=top["result"]["page"]["slug"],
            offer=offer["key"],
            query=buyer_intent,
            budget=buyer_budget,
            timeline="next 2 weeks",
            requested_terms={
                "location": buyer_location,
                "approvalBoundary": "Dry-run only. Do not contact seller until buyer approves.",
            },
        )
        recommended_next_step = "ask buyer to approve negotiation submission"
    else:
        validation = nexez.validate_checkout(
            slug=top["result"]["page"]["slug"],
            offer=offer["key"],
            query=buyer_intent,
            buyer_reference=f"location:{buyer_location}",
        )
        recommended_next_step = "ask buyer to approve checkout handoff"

    print(
        {
            "recommended_next_step": recommended_next_step,
            "selected_page": top["result"]["page"]["slug"],
            "selected_offer": offer["key"],
            "dry_run": validation,
        }
    )


def score_candidate(result: Dict[str, Any], manifest: Dict[str, Any]) -> int:
    offers = manifest.get("offers") or []
    location_signal = float((result.get("location_match") or {}).get("confidence") or 0) * 12
    action_signal = 10 if (result.get("offer") or {}).get("action") or any(item.get("action") or item.get("negotiation_action") for item in offers) else 0
    price_signal = 6 if (result.get("offer") or {}).get("price") or any(item.get("price") for item in offers) else 0
    faq_signal = 3 if manifest.get("faqs") else 0
    readiness = manifest.get("certification", {}).get("readiness", 0)
    readiness_signal = readiness / 10 if isinstance(readiness, (int, float)) else 0

    return round(float(result.get("score") or 0) + location_signal + action_signal + price_signal + faq_signal + readiness_signal)


if __name__ == "__main__":
    try:
        main()
    except NexezApiError as error:
        print({"error": str(error), "status": error.status, "body": error.body})

File: typescript/location-shortlist.ts
import { createNexezClient, type AgentPageManifest, type NexezSearchResult } from '@nexez/agent-sdk'

const nexez = createNexezClient({ buyerAgent: 'nexez-location-shortlist-example' })

const buyerIntent = process.env.NEXEZ_BUYER_INTENT ?? 'find a remote AI workflow consultant under 3000'
const buyerLocation = process.env.NEXEZ_BUYER_LOCATION ?? 'Chicago, IL'
const buyerBudget = process.env.NEXEZ_BUYER_BUDGET ?? 'USD 2500'

const matches = await nexez.search(buyerIntent, {
  location: buyerLocation,
  limit: 8,
})

const uniqueResults = [
  ...matches.results
    .reduce((pages, result) => {
      if (!pages.has(result.page.slug)) pages.set(result.page.slug, result)
      return pages
    }, new Map<string, NexezSearchResult>())
    .values(),
]
const candidates = await Promise.all(
  uniqueResults.slice(0, 5).map(async (result) => {
    const manifest = await nexez.getAgentPage(result.page.slug)
    const offerKey = result.offer?.key ?? manifest.offers[0]?.key
    const offer = offerKey ? manifest.offers.find((item) => item.key === offerKey) : undefined

    return {
      result,
      manifest,
      offer,
      score: scoreCandidate(result, manifest),
    }
  }),
)

const shortlist = candidates
  .filter((candidate) => candidate.offer)
  .sort((a, b) => b.score - a.score)
  .slice(0, 3)

console.log({
  query: buyerIntent,
  location: buyerLocation,
  shortlist: shortlist.map(({ result, manifest, offer, score }) => ({
    score,
    page: result.page.name,
    slug: result.page.slug,
    sellerLocation: result.page.location ?? manifest.page.location,
    offer: offer?.name,
    price: offer?.price,
    agentJson: result.page.agent_json_url,
  })),
})

const top = shortlist[0]
if (!top?.offer) {
  console.log('No actionable Nexez page found for this buyer intent.')
  process.exit(0)
}

const supportsNegotiation = Boolean(top.offer.negotiation_action)

const validation = supportsNegotiation
  ? await nexez.validateNegotiation({
      slug: top.result.page.slug,
      offer: top.offer.key,
      query: buyerIntent,
      budget: buyerBudget,
      timeline: 'next 2 weeks',
      requestedTerms: {
        location: buyerLocation,
        approvalBoundary: 'Dry-run only. Do not contact seller until buyer approves.',
      },
    })
  : await nexez.validateCheckout({
      slug: top.result.page.slug,
      offer: top.offer.key,
      query: buyerIntent,
      buyerReference: `location:${buyerLocation}`,
    })

console.log({
  recommendedNextStep: supportsNegotiation ? 'ask buyer to approve negotiation submission' : 'ask buyer to approve checkout handoff',
  selectedPage: top.result.page.slug,
  selectedOffer: top.offer.key,
  dryRun: validation,
})

function scoreCandidate(result: NexezSearchResult, manifest: AgentPageManifest): number {
  const locationSignal = (result.location_match?.confidence ?? 0) * 12
  const actionSignal = result.offer?.action || manifest.offers.some((offer) => offer.action || offer.negotiation_action) ? 10 : 0
  const priceSignal = result.offer?.price || manifest.offers.some((offer) => offer.price) ? 6 : 0
  const faqSignal = manifest.faqs?.length ? 3 : 0
  const readiness = readNumber(manifest.certification?.readiness)

  return Math.round((result.score || 0) + locationSignal + actionSignal + priceSignal + faqSignal + readiness / 10)
}

function readNumber(value: unknown): number {
  return typeof value === 'number' && Number.isFinite(value) ? value : 0
}
